spark_connect_labextension/handlers/base.py

```diff
@@ -0,0 +1,94 @@
+"""
+Base API Route Handler
+
+This file contains the base class inherited by all API route handlers, 
+which provides common property for accessing the extension configuration.
+"""
+
+import json
+from jupyter_server.base.handlers import APIHandler
+from jupyter_server.extension.handler import ExtensionHandlerMixin
+from jsonpath_ng import jsonpath, parse
+
+
+class SparkConnectAPIHandler(ExtensionHandlerMixin, APIHandler):
+    @property
+    def ext_config(self):
+        """
+        Property for retrieving extension config
+
+        :returns: extension config object
+        """
+        return self.settings['spark_connect_config']
+
+    @property
+    def spark_clusters(self):
+        """
+        Property for retrieving available clusters from configuration
+
+        :returns: array of cluster config object
+        """
+        return self.ext_config['clusters']
+
+    @property
+    def spark_config_bundles(self):
+        """
+        Property for retrieving Spark config bundles.
+        This function will merge the JSON object specified in `config_bundles`
+        and the one in the file referenced by `config_bundles_from_file` option.
+
+        :returns: dict of ConfigBundleName=ConfigBundleObject
+        """
+        if hasattr(self, '_spark_config_bundles'):
+            return self._spark_config_bundles
```

Comment:
Instead of manually memoizing the properties, note that you have `cached_property`:

https://docs.python.org/3/library/functools.html#functools.cached_property

Reply:
Okay will refactor this